File: DFSAndBFS/minDepth111.cc

/**
 * 111. Minimum Depth of Binary Tree
 * Given a binary tree, find its minimum depth.
 * The minimum depth is the number of nodes along the shortest path from the root node down to the nearest leaf node.
 * Note: A leaf is a node with no children.
 * */
#include<queue>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
 
class Solution {
public:
    int minDepth(TreeNode* root) {
        if (root == nullptr) return 0;
        queue<TreeNode*> nodeQueue;
        nodeQueue.push(root);
        int depth = 1;
        while (!nodeQueue.empty())
        {
            int n = nodeQueue.size();
            for(int i = 0; i< n; i++)
            {
                TreeNode *curNode = nodeQueue.front();
                nodeQueue.pop();
                if(curNode->left == nullptr && curNode->right == nullptr) return depth;
                if(curNode->left != nullptr) nodeQueue.push(curNode->left);
                if(curNode->right!= nullptr) nodeQueue.push(curNode->right);    
            }
            ++depth;
        }
        return depth;
    }

    /**
    int minDepth(TreeNode* root) {
        if(!root)
            return 0;
        int left=minDepth(root->left),right=minDepth(root->right);
        return (left && right) ? 1+min(left,right):1+left+right;
    }
     * */
};

File: DFSAndBFS/mindepth111/minDepth111.go
package mindepth111

//111. 二叉树的最小深度

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	queue := []*TreeNode{}
	queue = append(queue, root)
	depth := 1
	for len(queue) != 0 {
		sz := len(queue)
		for i := 0; i < sz; i++ {
			//从queue取出node
			cur := queue[0]
			queue = queue[1:]
			if cur.Left == nil && cur.Right == nil {
				return depth
			}
			if cur.Left != nil {
				queue = append(queue, cur.Left)
			}
			if cur.Right != nil {
				queue = append(queue, cur.Right)
			}
		}
		depth++
	}
	return depth
}
